File: Contato/app/src/main/java/br/com/imiranda/contato/ContatosAdapter.java
package br.com.imiranda.contato;

import android.content.Context;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;

import br.com.imiranda.contato.databinding.ViewContatoBinding;

public class ContatosAdapter extends ArrayAdapter<Contato> {
    public ContatosAdapter(Context context, int layout, ArrayList<Contato> contatosList){
        super(context,layout,contatosList);
    }

    @NonNull
    @Override
    public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
        ViewContatoBinding viewContatoBinding;
        ContatoViewHolder contatoViewHolder;
        //verificar se é necessário inflar(*criar uma nova célula)
        if (convertView == null) {
            viewContatoBinding = ViewContatoBinding.inflate(LayoutInflater.from(getContext()));
            //atribui a nova célula a view será devolvida preenchida para o Listview
            convertView = viewContatoBinding.getRoot();

            //pegar e guardar referências view interna
            contatoViewHolder = new ContatoViewHolder();
            contatoViewHolder.nomeContatoTV = viewContatoBinding.nomeContatoTv;
            contatoViewHolder.emailContatoTV = viewContatoBinding.emailContatoTv;
            //associa view celular
            convertView.setTag(contatoViewHolder);
        }
        //pegar o HOlder associado a célula (nova ou reciclada)
        contatoViewHolder = (ContatoViewHolder) convertView.getTag();


        Contato contato = getItem(position);
        contatoViewHolder.nomeContatoTV.setText(contato.getNome());
        contatoViewHolder.emailContatoTV.setText(contato.getEmail());

        return convertView;
    }
    private class ContatoViewHolder{
        public TextView nomeContatoTV;
        public TextView emailContatoTV;
    }
}

File: ContatosKt/app/src/main/java/br/com/imiranda/contatoskt/adapter/ContatosAdapter.kt
package br.com.imiranda.contatoskt.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import br.com.imiranda.contatoskt.R
import br.com.imiranda.contatoskt.model.Contato

class ContatosAdapter(private val contatosList: MutableList<Contato>,
                      private val onContatoClickListener: OnContatoClickListener
                      ):RecyclerView.Adapter<ContatosAdapter.ContatoViewHolder>() {
                          inner class ContatoViewHolder(viewContato: View): RecyclerView.ViewHolder(viewContato){
                              val nomeTv: TextView = viewContato.findViewById(R.id.nomeContatoTv)
                              val emailTv: TextView = viewContato.findViewById(R.id.emailContatoTv)
                          }
    //chamada pelo layout manager para criar uma nova view
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ContatoViewHolder {
        //val viewContatoBinding: ViewContatoBinding = ViewContatoBinding.inflate(LayoutInflater.from(parent.context))
        val viewContato: View = LayoutInflater.from(parent.context).inflate(R.layout.view_contato,parent,false)
        return ContatoViewHolder(viewContato)
        //return ContatoViewHolder(viewContatoBinding.root)
    }
    //chamada atualizar os valores de uma célula ou view
    override fun onBindViewHolder(holder: ContatoViewHolder, position: Int) {
        val contato: Contato = contatosList[position]
        holder.nomeTv.text = contato.nome
        holder.emailTv.text = contato.email
        holder.itemView.setOnClickListener{
            onContatoClickListener.onContatoClick(position)
        }
    }

//    override fun getItemCount(): Int {
//        return contatosList.size
//    }
    //single expression function
    override fun getItemCount(): Int = contatosList.size



}